### PR TITLE
Workaround bug in EvtFormatMessage

### LINKED ISSUE #100198: EventLogRecord.TaskDisplayName throws System.AccessViolationException and termintates the program
### Description

I am reading forwarded events (WEF) from event log. When I try to read the TaskDisplayName for an event where it does not exists application crashes with access violation exception. This also happens when reading the "KeywordsDisplayNames"

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at Microsoft.Win32.UnsafeNativeMethods.EvtFormatMessage(System.Diagnostics.Eventing.Reader.EventLogHandle, System.Diagnostics.Eventing.Reader.EventLogHandle, UInt32, Int32, EvtStringVariant[], EvtFormatMessageFlags, Int32, Char[], Int32 ByRef)
--------------------------------
   at System.Diagnostics.Eventing.Reader.NativeWrapper.EvtFormatMessageRenderName(System.Diagnostics.Eventing.Reader.EventLogHandle, System.Diagnostics.Eventing.Reader.EventLogHandle, EvtFormatMessageFlags)
   at System.Diagnostics.Eventing.Reader.ProviderMetadataCachedInformation.GetTaskDisplayName(System.String, System.Diagnostics.Eventing.Reader.EventLogHandle)
   at System.Diagnostics.Eventing.Reader.EventLogRecord.get_TaskDisplayName()
   at EventLogCrash.Program.Main(System.String[])

```

When I checked the event in the event viewer it's xml does not have values for Task and Keywords under RenderingInfo. 

```xml
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Microsoft-Windows-DNS-Client" Guid="{1C95126E-7EEA-49A9-A3FE-A378B03DDB4D}" />
    <EventID>1014</EventID>
    <Version>0</Version>
    <Level>3</Level>
    <Task>1014</Task>
    <Opcode>0</Opcode>
    <Keywords>0x4000000010000000</Keywords>
    <TimeCreated SystemTime="2024-03-21T23:32:09.577293500Z" />
    <EventRecordID>1590</EventRecordID>
    <Correlation />
    <Execution ProcessID="844" ThreadID="1280" />
    <Channel>System</Channel>
    <Computer>DESKTOP-JP7LKRA.kovo.local</Computer>
    <Security UserID="S-1-5-20" />
  </System>
  <EventData>
    <Data Name="QueryName">wpad</Data>
    <Data Name="AddressLength">128</Data>
    <Data Name="Address">02000000AC106480000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000</Data>
  </EventData>
  <RenderingInfo Culture="en-US">
    <Message>Name resolution for the name wpad timed out after none of the configured DNS servers responded.</Message>
    <Level>Warning</Level>
    <Task />
    <Opcode>Info</Opcode>
    <Channel>System</Channel>
    <Provider>Microsoft-Windows-DNS Client Events</Provider>
    <Keywords />
  </RenderingInfo>
</Event>
```

### Reproduction Steps

[EventLogCrash.zip](https://github.com/dotnet/runtime/files/14734158/EventLogCrash.zip)

The attached sample project contains an evtx file and the C# code tries to read the events from that file and it crashes with access violation.  

**Note:** Make sure to set the current machine's culture to en-US before running the sample. Otherwise it will not try to read the rendering info and wont fail. Setting the cultureinfo in c# code will not work. Changing of UI culture is not required

Reproducible on Windows 11 and Windows Server 2022. Have not checked for other operating systems.

### Expected behavior

The correct values should read and application should not crash.

### Actual behavior

The application crashes with Access Violation exception.  

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

- .NET 6.0.28 and 8.0.2
- Windows 11 and 2022
- x64
- I think it should happen to all operating systems. 

### Other information

_No response_

### PR DESCRIPTION
**Pull request by ericstj**
Fix #100198.  

This Windows API will still try to write to a buffer if you tell it the buffer is 0 length and the string you're reading happens to be 0 length.  We previously avoided an AV here because we were passing a StringBuilder, but that changed with the interop generator refactoring which passed `null`.  Avoid the bug that by passing in a 1-character buffer rather than null.

I've tested manually using the file provided in the issue repro.  I'll poke around a bit to see if I can manage to get a regression test for this without using the customer file but for now just sharing the fix.

### CODE REVIEW

File: src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs [hunk 1]
```diff
@@ -985,11 +987,12 @@ public static IEnumerable<string> EvtFormatMessageRenderKeywords(EventLogHandle
         {
             IntPtr buffer = IntPtr.Zero;
             int bufferNeeded;
+            char[] emptyBuffer = { '\0' }; // issue: https://github.com/dotnet/runtime/issues/100198
 
             try
             {
                 List<string> keywordsList = new List<string>();
-                bool status = UnsafeNativeMethods.EvtFormatMessageBuffer(pmHandle, eventHandle, 0, 0, IntPtr.Zero, flag, 0, IntPtr.Zero, out bufferNeeded);
```
Comment (carlossanlop) [inline src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs:992]: Why was this changed?

**Review by jkoritzinsky — APPROVED** [review]
Use collection expressions? Other than that, looks good.

**Review by carlossanlop — APPROVED** [review]
> I'll poke around a bit to see if I can manage to get a regression test for this without using the customer file but for now just sharing the fix.

Yes that would be great.

Otherwise, this LGTM.

File: src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs [hunk 2]
```diff
@@ -482,7 +482,8 @@ internal static EventLogHandle EvtGetPublisherMetadataPropertyHandle(EventLogHan
         public static string EvtFormatMessage(EventLogHandle handle, uint msgId)
         {
             int bufferNeeded;
-            bool status = UnsafeNativeMethods.EvtFormatMessage(handle, EventLogHandle.Zero, msgId, 0, null, UnsafeNativeMethods.EvtFormatMessageFlags.EvtFormatMessageId, 0, null, out bufferNeeded);
+            char[] emptyBuffer = { '\0' };  // issue: https://github.com/dotnet/runtime/issues/100198
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs:485]: This will allocate a new array on each call. Does this buffer get modified by EvtFormatMessage when the BufferSize is 0? If not, can it be a singleton array that's reused on every access?

Comment (ericstj) [reply src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs:485]: Yes, it's modified.  EvtFormatMessage will call memcopy to it, though in my debugging observations it's always copying zero (null character).  Because of this I was reluctant to use a singleton.  I also considered stackalloc, but felt it added a bit of complexity / unsafe code.  FWIW we previously allocated a StringBuilder here with a default array size of 16 characters.

File: src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs (hunk 1 above)
Comment (ericstj) [reply src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs:992]: The overload for `EvtFormatMessageBuffer` takes an `IntPtr` but now I'm calling the character version since I chose not to allocate the one character on the heap with `AllocHGlobal` as is done below.

I suppose this code is using `IntPtr` because it wants to just allocate the larger block of memory and parse things out with `PtrToStringUni` rather than slicing strings out of a character array.  Probably it's a side effect of previously this code used `StringBuilder` and that wouldn't handle pulling multiple strings out of a single buffer.  I think we could rewrite the below code to use a `char[]` and `slice` instead and save some use of `Marshal` but I'm not inclined to do that in this bugfix.

File: src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs (hunk 2 above)
Comment (ericstj) [reply src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/Reader/NativeWrapper.cs]: I was able to change to stackalloc, and I confirmed that changing the PInvoke to use `Span<char>` instead of `out char[]` didn't introduce any copy into the generated code.

**Review by stephentoub — APPROVED** [review]

**Comment by ericstj** [comment]
Corrected some mismatched active issues, filed a new active issue for a test crash in linq.expressions.  With that BA is passing.
